quill-jdbc-zio/src/main/scala/io/getquill/context/zio/ZioJdbcContext.scala

```diff
@@ -0,0 +1,287 @@
+    super.prepareQuery(sql, prepare, extractor)
+  override def prepareAction(sql: String, prepare: Prepare): Connection => Task[PreparedStatement] =
+    super.prepareAction(sql, prepare)
+  override def prepareBatchAction(groups: List[BatchGroup]): Connection => Task[List[PreparedStatement]] =
+    super.prepareBatchAction(groups)
+
+  override protected val effect: Runner = runner
+  import effect._
+
+  private val currentConnection: UIO[FiberRef[Option[Connection]]] = FiberRef.make(None)
+
+  override def close(): Unit = dataSource.close()
+
+  override protected def withConnection[T](f: Connection => Task[T]): Task[T] =
+    for {
+      maybeConnectionRef <- currentConnection
+      maybeConnection <- maybeConnectionRef.get
+      result <- maybeConnection match {
+        case Some(connection) => f(connection)
+        case None =>
+          schedule {
+            wrap(dataSource.getConnection).bracket(conn => wrapClose(conn.close()))(f)
+          }
+      }
+    } yield result
+
+  protected def withConnectionObservable[T](f: Connection => Stream[Throwable, T]): Stream[Throwable, T] =
+    for {
+      maybeConnectionRef <- Stream.fromEffect(currentConnection)
+      maybeConnection <- Stream.fromEffect(maybeConnectionRef.get)
+      result <- maybeConnection match {
+        case Some(connection) =>
+          withAutocommitBracket(connection, f)
+        case None =>
+          Stream.bracket(
+            Task(dataSource.getConnection)
+          )(conn => wrapClose(conn.close())).flatMap(conn => // Note: Can use mapM instead
+              withAutocommitBracket(conn, f))
+      }
+    } yield result
+
+  /**
+   * Need to store, set and restore the client's autocommit mode since some vendors (e.g. postgres)
+   * don't like autocommit=true during streaming sessions. Using brackets to do that.
+   */
+  private[getquill] def withAutocommitBracket[T](conn: Connection, f: Connection => Stream[Throwable, T]): Stream[Throwable, T] = {
+    Stream.bracket(Task(autocommitOff(conn)))(autoCommitBackOn)
+      .flatMap({ case (conn, _) => f(conn) })
+  }
+
+  private[getquill] def withAutocommitBracket[T](conn: Connection, f: Connection => Task[T]): Task[T] = {
+    Task(autocommitOff(conn))
+      .bracket(autoCommitBackOn)({ case (conn, _) => f(conn) })
+  }
+
+  private[getquill] def withCloseBracket[T](conn: Connection, f: Connection => Task[T]): Task[T] = {
+    Task(conn)
+      .bracket(conn => wrapClose(conn.close()))(conn => f(conn))
+  }
+
+  private[getquill] def autocommitOff(conn: Connection): (Connection, Boolean) = {
+    val ac = conn.getAutoCommit;
+    conn.setAutoCommit(false);
+    (conn, ac)
+  }
+
+  private[getquill] def autoCommitBackOn(state: (Connection, Boolean)) = {
+    val (conn, wasAutocommit) = state
+    wrapClose(conn.setAutoCommit(wasAutocommit))
+  }
+
+  def transaction[A](f: Task[A]): Task[A] = {
+    val dbEffects = for {
+      connectionRef <- currentConnection
```

Comment:
What do you think about ZIO.environment[Connection] instead of FiberRef?

Reply:
@adamgfraser suggested that too. Is we’re going in that direction, should `run` return a task with a Connection environment as well?

Reply:
Also, if we’re gonna do that, should we just rely on the user doing their own bracketing of the connection?

Reply:
And also, should we take environment of PreparedStatement in the cases that we use it. Or if we’re managing it ourselves should we bracket it?